core/domain/wipeout_service.py

```diff
@@ -0,0 +1,119 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Services for user data."""
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from constants import constants
+from core.domain import collection_services
+from core.domain import exp_fetchers
+from core.domain import exp_services
+from core.domain import skill_services
+from core.domain import topic_services
+from core.domain import user_services
+from core.platform import models
+
+current_user_services = models.Registry.import_current_user_services()
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+transaction_services = models.Registry.import_transaction_services()
+
+
+def _activity_should_be_deleted(activity_summary, user_id):
+    """Return if activity should be deleted, thus if it is private and the user
+    is the sole owner of the activity.
+
+    Args:
+        activity_summary: ExplorationSummary|CollectionSummary. The activity
+            summary.
+        user_id: str. The id of the user to be deleted.
+
+    Returns:
+        bool. Whether the activity should be deleted.
+    """
+    return (
+        activity_summary.status == constants.ACTIVITY_STATUS_PRIVATE and
+        user_id in activity_summary.owner_ids and
+        len(activity_summary.owner_ids) == 1
+    )
+
+
+def _topic_should_be_deleted(topic_rights, user_id):
+    """Return if topic should be deleted, thus if it is not published and
+    the user is the sole owner of the topic.
+
+    Args:
+        topic_rights: TopicRights. The topic rights.
+        user_id: str. The id of the user to be deleted.
+
+    Returns:
+        bool. Whether the topic should be deleted.
+    """
+    return (
+        not topic_rights.topic_is_published and
+        topic_rights.is_manager(user_id) and
+        len(topic_rights.manager_ids) == 1
+    )
+
+
+def pre_delete_user(user_id):
+    """Prepare user for the full deletion
+
+        Args:
+            user_id: str. The id of the user to be deleted.
+        """
+    subscribed_exploration_summaries = (
+        exp_fetchers.get_exploration_summaries_subscribed_to(user_id))
+    explorations_to_be_deleted_ids = [
+        exp_summary.id for exp_summary in subscribed_exploration_summaries
+        if _activity_should_be_deleted(exp_summary, user_id)]
+    for exp_id in explorations_to_be_deleted_ids:
+        exp_services.delete_exploration(user_id, exp_id)
```

Comment:
This is inefficient. Collect the IDs and do an ndb.delete_multi() at the end.

Reply:
Ah OK. But you should still be able to write a delete_multi() version of exp_services.delete_exploration(), right? We can deprecate the latter and use the new delete_multi() implementation instead (or, perhaps better, keep the delete_exploration() function but have it delegate to delete_multi()).

Reply:
I took a look at the code and it would require major refactoring, it would be probably possible to implement this for the `force_deletion`, but for the regular one it is quite a lot of models to handle. I would prefer to do the refactor in a separate PR if needed.

Reply:
I'm fine with tackling this in a separate PR -- please do the usual thing and file an issue and self-assign it, and take this on next. Otherwise this code is going to break when a user has lots of explorations. 

More generally, do we have a check somewhere in the process to verify that everything that should be deleted has been deleted? Just in case the deletion process fails midway for some GAE reason.